### ISSUE TITLE
Cilium fqdn/dnsproxy fails with FQDN dns policy when DNS respons is truncated

### ISSUE THREAD

**Issue by toredash** [cilium/cilium#31197]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?
https://github.com/cilium/cilium/pull/29239 introduced a new default which causes issues in certain setup

An egress.toFQDNs policy does not work under certain conditions when the DNS response is truncated and the host has IPv6 enabled in a IPv4-only environment. 

The dns proxy in cilium seems to be unable to connect to upstream coredns, but I don't think this is the real issue:
```
cilium-agent level=error msg="Cannot forward proxied DNS lookup" DNSRequestID=36304 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=523 error="failed to dial connection to 10.254.61.105:53: dial tcp 10.254.93.208:55534->10.254.61.105:53: connect: cannot assign requested address" identity=41536 ipAddr="10.254.93.208:55534" subsys=fqdn/dnsproxy
```

Setting dnsproxy-enable-transparent-mode to false corrects the issue.

There seem to be a related issue (https://github.com/cilium/cilium/issues/30669) where this change has caused issues as well.

I'm not sure what's specific to our networking layout that would cause this new default to create issues. I do note however that this specific comment seem to match our issue:
https://github.com/cilium/cilium/pull/28928#issuecomment-1788951996
```dnsproxy gets timeouts waiting for response from the dns server when curl is issuing both IPv4 (A) and IPv6 (AAAA) queries at the same time. Works flawlessly if limiting to IPv4 only (curl -4 ...).```

The issue seems to occur only if the DNS response has the DNS TC flag set.

This occurs only for this specific DNS query, any other queries pass through the proxy without issue or error msg. Here is the applied NetworkPolicy for the affected pod:

```yaml
---
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: test
spec:
  endpointSelector:
    matchLabels:
      test: "true"
  egress:
    - toFQDNs:
        - matchName: aoai-digital-pippen.openai.azure.com.
```   

With this policy present, I'm expecting that traffic to this endpoint is allowed:
```
# getent hosts aoai-digital-pippen.openai.azure.com.
20.61.99.32     api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com aoai-digital-pippen.openai.azure.com westeurope.api.cognitive.microsoft.com cognitiveweprod.trafficmanager.net cognitiveweprod.azure-api.net apimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net cognitiveweprod-westeurope-01.regional.azure-api.net
/# curl -v aoai-digital-pippen.openai.azure.com.
* Could not resolve host: aoai-digital-pippen.openai.azure.com.
* Closing connection 0
curl: (6) Could not resolve host: aoai-digital-pippen.openai.azure.com.
```

Putting cilium in debug mode, I was able to grab what fqdn/dnsproxy is logging related to this:
```
level=debug msg="Waited for endpoints to regenerate due to a DNS response" duration="70.817µs" endpointID=2638 qname=aoai-digital-pippen.openai.azure.com. subsys=daemon
level=debug msg="Responding to original DNS query" DNSRequestID=54559 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=2638 identity=29739 ipAddr="10.254.150.167:37343" subsys=fqdn/dnsproxy
level=debug msg="Received DNS response to proxied lookup" DNSRequestID=43037 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=2638 identity=29739 ipAddr="10.254.150.167:37343" resp=";; opcode: QUERY, status: NOERROR, id: 49148\n;; flags: qr tc rd ra; QUERY: 1, ANSWER: 5, AUTHORITY: 0, ADDITIONAL: 0\n\n;; QUESTION SECTION:\n;aoai-digital-pippen.openai.azure.com.\tIN\t AAAA\n\n;; ANSWER SECTION:\naoai-digital-pippen.openai.azure.com.\t5\tIN\tCNAME\twesteurope.api.cognitive.microsoft.com.\nwesteurope.api.cognitive.microsoft.com.\t5\tIN\tCNAME\tcognitiveweprod.trafficmanager.net.\ncognitiveweprod.trafficmanager.net.\t5\tIN\tCNAME\tcognitiveweprod.azure-api.net.\ncognitiveweprod.azure-api.net.\t5\tIN\tCNAME\tapimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net.\napimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net.\t5\tIN\tCNAME\tcognitiveweprod-westeurope-01.regional.azure-api.net.\n" subsys=fqdn/dnsproxy
level=debug msg="Notifying with DNS response to original DNS query" DNSRequestID=43037 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=2638 identity=29739 ipAddr="10.254.150.167:37343" subsys=fqdn/dnsproxy
level=debug msg="Responding to original DNS query" DNSRequestID=43037 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=2638 identity=29739 ipAddr="10.254.150.167:37343" subsys=fqdn/dnsproxy
level=debug msg="Handling DNS query from endpoint" DNSRequestID=54559 dnsName=aoai-digital-pippen.openai.azure.com. ipAddr="10.254.150.167:57560" subsys=fqdn/dnsproxy
level=debug msg="Found target server to of DNS request secID {ID:10023 Source:custom-resource _:[] shadowed:false createdFromMetadata:false}" DNSRequestID=54559 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=2638 identity=29739 ipAddr="10.254.150.167:57560" server="10.254.87.63:53" subsys=fqdn/dnsproxy
level=debug msg="Forwarding DNS request for a name that is allowed" DNSRequestID=54559 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=2638 identity=29739 ipAddr="10.254.150.167:57560" subsys=fqdn/dnsproxy
level=debug msg="Received DNS response to proxied lookup" DNSRequestID=54559 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=2638 identity=29739 ipAddr="10.254.150.167:57560" resp=";; opcode: QUERY, status: NOERROR, id: 44857\n;; flags: qr aa rd ra; QUERY: 1, ANSWER: 7, AUTHORITY: 0, ADDITIONAL: 0\n\n;; QUESTION SECTION:\n;aoai-digital-pippen.openai.azure.com.\tIN\t A\n\n;; ANSWER SECTION:\naoai-digital-pippen.openai.azure.com.\t2\tIN\tCNAME\twesteurope.api.cognitive.microsoft.com.\nwesteurope.api.cognitive.microsoft.com.\t2\tIN\tCNAME\tcognitiveweprod.trafficmanager.net.\ncognitiveweprod.trafficmanager.net.\t2\tIN\tCNAME\tcognitiveweprod.azure-api.net.\ncognitiveweprod.azure-api.net.\t2\tIN\tCNAME\tapimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net.\napimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net.\t2\tIN\tCNAME\tcognitiveweprod-westeurope-01.regional.azure-api.net.\ncognitiveweprod-westeurope-01.regional.azure-api.net.\t2\tIN\tCNAME\tapi0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com.\napi0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com.\t2\tIN\tA\t20.61.99.32\n" subsys=fqdn/dnsproxy
level=debug msg="Notifying with DNS response to original DNS query" DNSRequestID=54559 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=2638 identity=29739 ipAddr="10.254.150.167:57560" subsys=fqdn/dnsproxy
level=debug msg="Updating DNS name in cache from response to query" ips="[20.61.99.32]" qname=aoai-digital-pippen.openai.azure.com. subsys=daemon
level=debug msg="FQDN: IPs didn't change for DNS name" dnsName=aoai-digital-pippen.openai.azure.com. lookupIPs="&{2 [20.61.99.32]}" subsys=fqdn
level=debug msg="Waited for endpoints to regenerate due to a DNS response" duration="53.699µs" endpointID=2638 qname=aoai-digital-pippen.openai.azure.com. subsys=daemon
level=debug msg="Responding to original DNS query" DNSRequestID=54559 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=2638 identity=29739 ipAddr="10.254.150.167:57560" subsys=fqdn/dnsproxy
level=debug msg="Handling DNS query from endpoint" DNSRequestID=43037 dnsName=aoai-digital-pippen.openai.azure.com. ipAddr="10.254.150.167:57560" subsys=fqdn/dnsproxy
level=debug msg="Found target server to of DNS request secID {ID:10023 Source:custom-resource _:[] shadowed:false createdFromMetadata:false}" DNSRequestID=43037 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=2638 identity=29739 ipAddr="10.254.150.167:57560" server="10.254.87.63:53" subsys=fqdn/dnsproxy
level=debug msg="Forwarding DNS request for a name that is allowed" DNSRequestID=43037 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=2638 identity=29739 ipAddr="10.254.150.167:57560" subsys=fqdn/dnsproxy
level=error msg="Cannot forward proxied DNS lookup" DNSRequestID=43037 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=2638 error="failed to dial connection to 10.254.87.63:53: dial tcp 10.254.150.167:57560->10.254.87.63:53: connect: cannot assign requested address" identity=29739 ipAddr="10.254.150.167:57560" subsys=fqdn/dnsproxy
```

Notice that the _first_ DNS query returns an incomplete response (see `ANSWER: 5, `), but with the `tc` flag set, so the response is in-complete as it does not fit into the UDP packet. It then re-queries over TCP, (see `ANSWER: 7, `), and the complete query is returned.

For reference, this is the answer payload I saw locally when testing this:
```shell
;; ANSWER SECTION:
aoai-digital-pippen.openai.azure.com. 900 IN CNAME westeurope.api.cognitive.microsoft.com.
westeurope.api.cognitive.microsoft.com.	3600 IN	CNAME cognitiveweprod.trafficmanager.net.
cognitiveweprod.trafficmanager.net. 30 IN CNAME	cognitiveweprod.azure-api.net.
cognitiveweprod.azure-api.net. 900 IN	CNAME	apimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net.
apimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net. 300 IN CNAME cognitiveweprod-westeurope-01.regional.azure-api.net.
cognitiveweprod-westeurope-01.regional.azure-api.net. 900 IN CNAME api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com.
api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com.	10 IN A	20.61.99.32
```

Within the affected pod, DNS query works as expected:
```shell
# getent hosts aoai-digital-pippen.openai.azure.com.
20.61.99.32     api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com aoai-digital-pippen.openai.azure.com westeurope.api.cognitive.microsoft.com cognitiveweprod.trafficmanager.net cognitiveweprod.azure-api.net apimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net cognitiveweprod-westeurope-01.regional.azure-api.net
/# cat /etc/resolv.conf 
search redact.svc.cluster.local svc.cluster.local cluster.local eu-north-1.compute.internal
nameserver 10.255.0.10
options ndots:5
```

While doing the lookup, the following Hubble flow is observed:
```yaml
redact:41566 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 EGRESS ALLOWED (UDP)
redact:41566 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (UDP)
redact:41566 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) dns-request proxy FORWARDED (DNS Query aoai-digital-pippen.openai.azure.com. AAAA)
redact:41566 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 INGRESS ALLOWED (UDP)
redact:41566 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (UDP)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 EGRESS ALLOWED (TCP Flags: SYN)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: SYN)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK, PSH)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) dns-request proxy FORWARDED (DNS Query aoai-digital-pippen.openai.azure.com. AAAA)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 INGRESS ALLOWED (TCP Flags: SYN)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: SYN)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: ACK)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK, FIN)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: ACK)
redact:39642 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK)
redact:51481 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 EGRESS ALLOWED (UDP)
redact:51481 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (UDP)
redact:51481 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) dns-request proxy FORWARDED (DNS Query aoai-digital-pippen.openai.azure.com. A)
redact:51481 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 INGRESS ALLOWED (UDP)
redact:51481 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (UDP)
redact:54074 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) policy-verdict:L3-L4 EGRESS ALLOWED (TCP Flags: SYN)
redact:54074 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) to-proxy FORWARDED (TCP Flags: SYN)
redact:54074 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK)
redact:54074 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK, PSH)
redact:54074 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) dns-request proxy FORWARDED (DNS Query aoai-digital-pippen.openai.azure.com. A)
redact:54074 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK, FIN)
redact:54074 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK)
```

On the other hand, curl will fail even though the policy should allow the traffic:
```shell
# curl aoai-digital-pippen.openai.azure.com. -v
* Could not resolve host: aoai-digital-pippen.openai.azure.com.
* Closing connection 0
curl: (6) Could not resolve host: aoai-digital-pippen.openai.azure.com.
```

This is when cilium will log the following error:
```
level=error msg="Cannot forward proxied DNS lookup" DNSRequestID=52940 dnsName=aoai-digital-pippen.openai.azure.com. endpointID=523 error="failed to dial connection to 10.254.61.105:53: dial tcp 10.254.93.208:59448->10.254.61.105:53: connect: cannot assign requested address" identity=41536 ipAddr="10.254.93.208:59448" subsys=fqdn/dnsproxy 
```

I doubt this is an issue with enough local ports, as the error is only generated for that specific query in our environment.

Hubble flows for this curl command shows an PROXY ERROR for the AAAA record, bot no error for the A record:
```redact:34083 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 EGRESS ALLOWED (UDP)
redact:34083 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (UDP)
redact:34083 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) dns-request proxy FORWARDED (DNS Query aoai-digital-pippen.openai.azure.com. AAAA)
redact:34083 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) dns-request proxy FORWARDED (DNS Query aoai-digital-pippen.openai.azure.com. A)
redact:34083 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 INGRESS ALLOWED (UDP)
redact:34083 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (UDP)
redact:59448 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) policy-verdict:L3-L4 EGRESS ALLOWED (TCP Flags: SYN)
redact:59448 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) to-proxy FORWARDED (TCP Flags: SYN)
redact:59448 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK)
redact:59448 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK, PSH)
redact:59448 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) dns-request proxy FORWARDED (DNS Query aoai-digital-pippen.openai.azure.com. A)
redact:59448 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) dns-request proxy FORWARDED (DNS Query aoai-digital-pippen.openai.azure.com. AAAA)
redact:59448 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) dns-request proxy ERROR (DNS Query aoai-digital-pippen.openai.azure.com. AAAA)
redact:59448 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK, FIN)
```

For some reason, the curl command will work if one specify it with `-4` (This option tells curl to use IPv4 addresses only when resolving hostnames, and not for example try IPv6.):
```
# curl -4 aoai-digital-pippen.openai.azure.com. -I  --silent 
HTTP/1.1 400 Bad Request
Content-Length: 334
Content-Type: text/html; charset=us-ascii
Date: Wed, 06 Mar 2024 11:42:01 GMT
Connection: close
```

Cilium agent does not report any errors, and Hubble flow shows no error, but also no AAAA queries:
```
redact:46155 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) policy-verdict:L3-L4 EGRESS ALLOWED (UDP)
redact:46155 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) to-proxy FORWARDED (UDP)
redact:46155 (ID:41536) -> kube-system/coredns-7f6fdfd5c-p5tph:53 (ID:10023) dns-request proxy FORWARDED (DNS Query aoai-digital-pippen.openai.azure.com. A)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 EGRESS ALLOWED (TCP Flags: SYN)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: SYN)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK, PSH)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) dns-request proxy FORWARDED (DNS Query aoai-digital-pippen.openai.azure.com. A)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 INGRESS ALLOWED (TCP Flags: SYN)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: SYN)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: ACK)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK, FIN)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK)
redact:54172 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: ACK)
redact:48856 (ID:41536) -> aoai-digital-pippen.openai.azure.com:80 (ID:16778731) policy-verdict:L3-L4 EGRESS ALLOWED (TCP Flags: SYN)
redact:48856 (ID:41536) -> aoai-digital-pippen.openai.azure.com:80 (ID:16778731) to-network FORWARDED (TCP Flags: SYN)
redact:48856 (ID:41536) -> aoai-digital-pippen.openai.azure.com:80 (ID:16778731) to-network FORWARDED (TCP Flags: ACK)
redact:48856 (ID:41536) -> aoai-digital-pippen.openai.azure.com:80 (ID:16778731) to-network FORWARDED (TCP Flags: ACK, PSH)
redact:48856 (ID:41536) -> aoai-digital-pippen.openai.azure.com:80 (ID:16778731) to-network FORWARDED (TCP Flags: ACK, FIN)
```

I'm not sure what's wrong here. If I update resolv.conf to include the [no-aaaa](https://man7.org/linux/man-pages/man5/resolv.conf.5.html) option, the problem goes away. But I find it strange and not expected that a failed AAAA query should deny traffic in this fashion.
```
resolv.conf :  this works
search redact.svc.cluster.local svc.cluster.local cluster.local eu-north-1.compute.internal.cluster.local cluster.local eu-north-1.compute.internal
nameserver 10.255.0.10
options ndots:5 no-aaaa

# curl aoai-digital-pippen.openai.azure.com.
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd">
<HTML><HEAD><TITLE>Bad Request</TITLE>
<META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD>
<BODY><h2>Bad Request - Invalid Hostname</h2>
<hr><p>HTTP Error 400. The request hostname is invalid.</p>
</BODY></HTML>

redact:40135 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 EGRESS ALLOWED (UDP)
redact:40135 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (UDP)
redact:40135 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) dns-request proxy FORWARDED (DNS Query aoai-digital-pippen.openai.azure.com. A)
redact:40135 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 INGRESS ALLOWED (UDP)
redact:40135 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (UDP)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 EGRESS ALLOWED (TCP Flags: SYN)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: SYN)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK, PSH)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) dns-request proxy FORWARDED (DNS Query aoai-digital-pippen.openai.azure.com. A)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) policy-verdict:L3-L4 INGRESS ALLOWED (TCP Flags: SYN)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: SYN)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: ACK)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: ACK, PSH)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: ACK, FIN)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK, FIN)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-proxy FORWARDED (TCP Flags: ACK)
redact:53652 (ID:41536) -> kube-system/coredns-7f6fdfd5c-fccrw:53 (ID:10023) to-endpoint FORWARDED (TCP Flags: ACK)
redact:40056 (ID:41536) -> aoai-digital-pippen.openai.azure.com:80 (ID:16778909) policy-verdict:L3-L4 EGRESS ALLOWED (TCP Flags: SYN)
redact:40056 (ID:41536) -> aoai-digital-pippen.openai.azure.com:80 (ID:16778909) to-network FORWARDED (TCP Flags: SYN)
redact:40056 (ID:41536) -> aoai-digital-pippen.openai.azure.com:80 (ID:16778909) to-network FORWARDED (TCP Flags: ACK)
redact:40056 (ID:41536) -> aoai-digital-pippen.openai.azure.com:80 (ID:16778909) to-network FORWARDED (TCP Flags: ACK, PSH)
redact:40056 (ID:41536) -> aoai-digital-pippen.openai.azure.com:80 (ID:16778909) to-network FORWARDED (TCP Flags: ACK, FIN)
```

I have a hunch that if there is something wrong with the fqdn/proxy component. Does it handle an incomplete query response for AAAA records properly ? As `aoai-digital-pippen.openai.azure.com.` for AAAA is just a CNAME chain that does not result in something:
```
 dig AAAA aoai-digital-pippen.openai.azure.com @8.8.8.8

; <<>> DiG 9.10.6 <<>> AAAA aoai-digital-pippen.openai.azure.com @8.8.8.8
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 40269
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;aoai-digital-pippen.openai.azure.com. IN AAAA

;; ANSWER SECTION:
aoai-digital-pippen.openai.azure.com. 900 IN CNAME westeurope.api.cognitive.microsoft.com.
westeurope.api.cognitive.microsoft.com.	1134 IN	CNAME cognitiveweprod.trafficmanager.net.
cognitiveweprod.trafficmanager.net. 15 IN CNAME	cognitiveweprod.azure-api.net.
cognitiveweprod.azure-api.net. 855 IN	CNAME	apimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net.
apimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net. 300 IN CNAME cognitiveweprod-westeurope-01.regional.azure-api.net.
cognitiveweprod-westeurope-01.regional.azure-api.net. 713 IN CNAME api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com.

;; AUTHORITY SECTION:
westeurope.cloudapp.azure.com. 60 IN	SOA	ns1-201.azure-dns.com. msnhst.microsoft.com. 10001 900 300 604800 60

;; Query time: 133 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Wed Mar 06 13:56:40 CET 2024
;; MSG SIZE  rcvd: 472
```

I've briefly looked at the [proxy code](https://github.com/cilium/cilium/blob/f92b528abc30a5c66ba80d5922d4e58c48cfe7e1/pkg/fqdn/dnsproxy/proxy.go#L1135) that relates to CNAMES and AAAA, but I'm not able to identify if something special happens when a AAAA query returns a set of CNAMES that does not resolve to anything.

IPv6 is not disabled in the kernel.

### Cilium Version

# cilium version
Client: 1.15.1 a368c8f0 2024-02-14T22:16:57+00:00 go version go1.21.6 linux/amd64
Daemon: 1.15.1 a368c8f0 2024-02-14T22:16:57+00:00 go version go1.21.6 linux/amd64

### Kernel Version

# uname -a
Linux ip-10-209-136-25.eu-north-1.compute.internal 5.10.209-198.858.amzn2.x86_64 #1 SMP Tue Feb 13 18:46:41 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Server Version: v1.28.5-eks-5e0fdde

### Relevant log output

```shell
apiVersion: v1
data:
  agent-not-ready-taint-key: node.cilium.io/agent-not-ready
  arping-refresh-period: 30s
  auto-create-cilium-node-resource: "true"
  auto-direct-node-routes: "false"
  aws-enable-prefix-delegation: "true"
  bpf-lb-acceleration: disabled
  bpf-lb-external-clusterip: "false"
  bpf-lb-map-max: "65536"
  bpf-lb-sock: "false"
  bpf-map-dynamic-size-ratio: "0.0025"
  bpf-policy-map-max: "16384"
  bpf-root: /sys/fs/bpf
  cgroup-root: /run/cilium/cgroupv2
  cilium-endpoint-gc-interval: 5m0s
  cluster-id: "0"
  cluster-name: default
  cni-exclusive: "true"
  cni-log-file: /var/run/cilium/cilium-cni.log
  controller-group-metrics: write-cni-file sync-host-ips sync-lb-maps-with-k8s-services
  custom-cni-conf: "false"
  debug: "false"
  debug-verbose: ""
  devices: eth0
  dnsproxy-enable-transparent-mode: "true"
  ec2-api-endpoint: ""
  egress-gateway-reconciliation-trigger-interval: 1s
  enable-auto-protect-node-port-range: "true"
  enable-bgp-control-plane: "false"
  enable-bpf-clock-probe: "false"
  enable-bpf-masquerade: "true"
  enable-endpoint-health-checking: "true"
  enable-endpoint-routes: "true"
  enable-health-check-loadbalancer-ip: "false"
  enable-health-check-nodeport: "true"
  enable-health-checking: "true"
  enable-hubble: "true"
  enable-hubble-open-metrics: "false"
  enable-ip-masq-agent: "true"
  enable-ipv4: "true"
  enable-ipv4-big-tcp: "false"
  enable-ipv4-masquerade: "true"
  enable-ipv6: "false"
  enable-ipv6-big-tcp: "false"
  enable-ipv6-masquerade: "true"
  enable-k8s-networkpolicy: "true"
  enable-k8s-terminating-endpoint: "true"
  enable-l2-neigh-discovery: "true"
  enable-l7-proxy: "true"
  enable-local-redirect-policy: "false"
  enable-masquerade-to-route-source: "false"
  enable-metrics: "true"
  enable-policy: default
  enable-remote-node-identity: "true"
  enable-runtime-device-detection: "true"
  enable-sctp: "false"
  enable-svc-source-range-check: "true"
  enable-vtep: "false"
  enable-well-known-identities: "false"
  enable-xt-socket-fallback: "true"
  eni-tags: '{}'
  external-envoy-proxy: "false"
  hubble-disable-tls: "false"
  hubble-export-file-max-backups: "5"
  hubble-export-file-max-size-mb: "10"
  hubble-listen-address: :4244
  hubble-metrics: dns:query;ignoreAAAA drop:sourceContext=workload-name|pod-name|ip;destinationContext=workload-name|pod-name|dns|ip;labelsContext=source_namespace,destination_namespace
    tcp flow:sourceContext=workload-name|pod-name|ip;destinationContext=workload-name|pod-name|dns|ip;labelsContext=source_namespace,destination_namespace
    icmp http
  hubble-metrics-server: :9965
  hubble-socket-path: /var/run/cilium/hubble.sock
  hubble-tls-cert-file: /var/lib/cilium/tls/hubble/server.crt
  hubble-tls-client-ca-files: /var/lib/cilium/tls/hubble/client-ca.crt
  hubble-tls-key-file: /var/lib/cilium/tls/hubble/server.key
  identity-allocation-mode: crd
  identity-gc-interval: 15m0s
  identity-heartbeat-timeout: 30m0s
  install-no-conntrack-iptables-rules: "false"
  instance-tags-filter: kubernetes.io/cluster/ip-eks-stage=owned
  ipam: eni
  ipam-cilium-node-update-rate: 15s
  k8s-client-burst: "20"
  k8s-client-qps: "10"
  kube-proxy-replacement: "true"
  kube-proxy-replacement-healthz-bind-address: ""
  max-connected-clusters: "255"
  mesh-auth-enabled: "true"
  mesh-auth-gc-interval: 5m0s
  mesh-auth-queue-size: "1024"
  mesh-auth-rotated-identities-queue-size: "1024"
  monitor-aggregation: medium
  monitor-aggregation-flags: all
  monitor-aggregation-interval: 5s
  node-port-bind-protection: "true"
  nodes-gc-interval: 5m0s
  operator-api-serve-addr: 127.0.0.1:9234
  operator-prometheus-serve-addr: :9963
  policy-audit-mode: "false"
  policy-cidr-match-mode: ""
  preallocate-bpf-maps: "false"
  procfs: /host/proc
  prometheus-serve-addr: :9962
  proxy-connect-timeout: "2"
  proxy-max-connection-duration-seconds: "0"
  proxy-max-requests-per-connection: "0"
  proxy-prometheus-port: "9964"
  remove-cilium-node-taints: "true"
  routing-mode: native
  service-no-backend-response: reject
  set-cilium-is-up-condition: "true"
  set-cilium-node-taints: "true"
  sidecar-istio-proxy-image: cilium/istio_proxy
  skip-cnp-status-startup-clean: "false"
  subnet-ids-filter: subnet-0b1aa525788de8a9c subnet-0449b556e3d2e0689 subnet-09b9e3a77f7db3b32
  synchronize-k8s-nodes: "true"
  tofqdns-dns-reject-response-code: refused
  tofqdns-enable-dns-compression: "true"
  tofqdns-endpoint-max-ip-per-hostname: "50"
  tofqdns-idle-connection-grace-period: 0s
  tofqdns-max-deferred-connection-deletes: "10000"
  tofqdns-proxy-response-max-delay: 100ms
  unmanaged-pod-watcher-interval: "15"
  update-ec2-adapter-limit-via-api: "true"
  vtep-cidr: ""
  vtep-endpoint: ""
  vtep-mac: ""
  vtep-mask: ""
  write-cni-conf-when-ready: /host/etc/cni/net.d/05-cilium.conflist
```

### Cilium Users Document

- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by toredash** [comment]
Hi @joestringer ,

I believe the reason for the workload to issue IPv6, as well IPv4 DNS request, is not the apps fault, but rather glibc. Note the client is issuing AAAA and A lookup _over_ IPv4, not IPv6. IPv6 is not involved here in the transports of the packets over the wire. At some point in glibc, i believe they started to do queries for both A and AAAA records. One can avoid this if one configures resolv.conf with [options no-aaaa](https://man7.org/linux/man-pages/man5/resolv.conf.5.html), or, if getaddrinfo() is called with [ai_family = AF_INET](https://man7.org/linux/man-pages/man3/getaddrinfo.3.html) instead of the default value of `AF_UNSPEC`

The glibc for the affected pod ([python:3.9](https://hub.docker.com/_/python)):
```
# ldconfig -V
ldconfig (Debian GLIBC 2.36-9+deb12u4) 2.36
````

The EKSI AMI, when booted in EC2, comes with a link loca (fe80::/10) address, but other than that there is no IPv6 in this environment. The subnets are IPv4 only, no NAT64 or DNS64 is deployed in this environment. There should be no reason to do any AAAA lookups in this environment, and one fix is of course to add `no-aaaa` to resolv.conf, but is seems there is another issue here that could be fixed.

In glibc 2.9, it seems they introduced parallel queries for A and AAAA, and in glibc2.10 they introduced they introduced [single-request in resolv.conf](https://man7.org/linux/man-pages/man5/resolv.conf.5.html) to not send them in parallel. The comment in https://github.com/cilium/cilium/pull/28928#issuecomment-1788951996 just seems to fit right in there, but I have a hard time pinpointing it.

The issue also seems to only trigger if the AAAA response is truncated, while the A response is not. The hunch there is that the client might be in limbo; as the A record is "complete" and returns an actual IP, the AAAA record is kinda complete, but with the TC flag set, to instruct the client to re-query over TCP. 

Would it help with a tcpdump of the issue, one with dnsproxy-enable-transparent-mode to false and one with true ?

If needed I can look at additional testing with different container images, do you have a preferred image that I should use?

**Comment by joestringer** [comment]
Ah OK, I missed the forest for the trees in my initial assessment. I see what you mean about the traffic being entirely IPv4 and that the only IPv6 aspect is specifically the AAAA query+response which is transmitted over IPv4.

> The issue also seems to only trigger if the AAAA response is truncated, while the A response is not.

Are you able to demonstrate why the AAAA responses are truncated? Is this because the resulting records are larger than MTU and so the UDP DNS response cannot fit in a single packet, or is there something else going on?

> Would it help with a tcpdump of the issue, one with dnsproxy-enable-transparent-mode to false and one with true ?

If this is the difference between the working and non-working case, then yes I think this would be very helpful.

In the mean time I've triaged this to sig-agent in the proxy area so folks can more readily find and investigate this issue deeper.

**Comment by toredash** [comment]
> Are you able to demonstrate why the AAAA responses are truncated? Is this because the resulting records are larger than MTU and so the UDP DNS response cannot fit in a single packet, or is there something else going on?

The AAAA response is truncated as the response does not fit an UDP packet. From the tcpdump, the dns.flags.truncated flag is set to True.

@joestringer I've been able to pinpoint this and with a reproducible method.

The issue seems to be only affecting glibc based workloads. musl based workloads are not affected.

**All** following items must be true for the issue to trigger:
* The glibc based workloads must issue queries for A and AAAA requests in parallel. 
* The AAAA response must be truncated
* dnsproxy-enable-transparent-mode must be set to true

To reproduce use the following manifests:
```yaml
apiVersion: cilium.io/v2
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: default-policy
spec:
  endpointSelector: {}
  ingress:
    - fromEndpoints:
        - {}
  egress:
    - toEndpoints:
        - {}  
    - toEndpoints:
        - matchLabels:
            io.kubernetes.pod.namespace: kube-system
            k8s-app: kube-dns
      toPorts:
        - ports:
            - port: "53"
              protocol: ANY
          rules:
            dns:
              - matchPattern: "*"
---
apiVersion: v1
kind: Pod
metadata:
  name: cilium-debugging-dns-alpine
spec:
  containers:
    - name: cilium-debugging-dns
      image: alpine:edge
      command: ["/bin/sh", "-c", "--"]
      args: ["while true; do sleep 3600; done;"]
---
apiVersion: v1
kind: Pod
metadata:
  name: cilium-debugging-dns-debian
spec:
  containers:
    - name: cilium-debugging-dns
      image: debian:bookworm
      command: ["/bin/sh", "-c", "--"]
      args: ["while true; do sleep 3600; done;"]
```

To reproduce one can do a lookup towards this record:
`westeurope.api.cognitive.microsoft.com.`

Notice that this record returns a long CNAME chain, here with a A query
```
westeurope.api.cognitive.microsoft.com.	3226 IN	CNAME cognitiveweprod.trafficmanager.net.
cognitiveweprod.trafficmanager.net. 26 IN CNAME	cognitiveweprod.azure-api.net.
cognitiveweprod.azure-api.net. 282 IN	CNAME	apimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net.
apimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net. 300 IN CNAME cognitiveweprod-westeurope-01.regional.azure-api.net.
cognitiveweprod-westeurope-01.regional.azure-api.net. 195 IN CNAME api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com.
api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com.	6 IN A 20.61.99.32
```

And the same with AAAA query:
```
;; ANSWER SECTION:
westeurope.api.cognitive.microsoft.com.	3219 IN	CNAME cognitiveweprod.trafficmanager.net.
cognitiveweprod.trafficmanager.net. 30 IN CNAME	cognitiveweprod.azure-api.net.
cognitiveweprod.azure-api.net. 234 IN	CNAME	apimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net.
apimgmttmdjylckcx6clmh2isu2wr38uqzm63s8n4ub2y3e6xs.trafficmanager.net. 300 IN CNAME cognitiveweprod-westeurope-01.regional.azure-api.net.
cognitiveweprod-westeurope-01.regional.azure-api.net. 234 IN CNAME api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com.
```

Note that the AAAA query does not return a AAAA record, as `api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com.` does not return a AAAA record.

In the ubuntu pod, issue the following command:
```shell
# while true; do getent ahosts westeurope.api.cognitive.microsoft.com. || echo "Did not work"; done
Did not work
Did not work
Did not work
20.61.99.32     STREAM api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com
20.61.99.32     DGRAM  
20.61.99.32     RAW    
20.61.99.32     STREAM api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com
20.61.99.32     DGRAM  
20.61.99.32     RAW    
Did not work
20.61.99.32     STREAM api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com
20.61.99.32     DGRAM  
20.61.99.32     RAW    
Did not work
Did not work
Did not work
```

It sometimes works, sometime it does not. When it does not work cilium logs the following:
```cilium-agent level=error msg="Cannot forward proxied DNS lookup" DNSRequestID=50342 dnsName=westeurope.api.cognitive.microsoft.com. endpointID=1221 error="failed to dial connection to 10.254.20.28:53: dial tcp 10.254.59.223:45826->10.254.20.28:53: connect: cannot assign requested address" identity=20939 ipAddr="10.254.59.223:45826" subsys=fqdn/dnsproxy
```

In the alpine pod, this is not reproducible:
```
# while true; do getent ahosts westeurope.api.cognitive.microsoft.com. || echo "Did not work"; done
20.61.99.32     STREAM api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com
20.61.99.32     DGRAM  
20.61.99.32     RAW    
20.61.99.32     STREAM api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com
20.61.99.32     DGRAM  
20.61.99.32     RAW    
20.61.99.32     STREAM api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com
20.61.99.32     DGRAM  
20.61.99.32     RAW    
20.61.99.32     STREAM api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com
20.61.99.32     DGRAM  
20.61.99.32     RAW    
```

While the testing was performed, no coredns pods was rolled. If one does a query for the last CNAME in the chain (`api0df69e61b7d2469bb25a53be6441fcf3pcvso5kj6wdvbahismwbl.westeurope.cloudapp.azure.com.`), there is not an issue for either the ubuntu or alpine pod.

The issue goes away if we change  dnsproxy-enable-transparent-mode from true to false 😕 

I don't have another environment to test on, but I hope this is enough information for you to try to reproduce it in another environment.

**Comment by kahirokunn** [comment]
I got the same error.

**Comment by kahirokunn** [comment]
The `getent ahosts registry.yarnpkg.com` command gives the same error.

**Comment by toredash** [comment]
> The `getent ahosts registry.yarnpkg.com` command gives the same error. However, the error does not occur with `getent ahostsv6`.

I can confirm that this lookup is also affected on glibc based workloads.

ı've attached two tcpdumps, `dns_yarn.pcap.gz` shows when the query fails, and `dns_yarn_working.pcap.gz` when it succeeds on the next  run of `getent ahosts registry.yarnpkg.com`.

[dns_yarn_working.pcap.gz](https://github.com/cilium/cilium/files/14556112/dns_yarn_working.pcap.gz)
[dns_yarn.pcap.gz](https://github.com/cilium/cilium/files/14556114/dns_yarn.pcap.gz)

**Comment by kahirokunn** [comment]
Keep

**Comment by toredash** [comment]
@joestringer anything else I can provide in terms of data or examples?

**Comment by DingGGu** [comment]
After I upgrade Cilium v1.14.5 to v1.14.11,
I found dropped packet in hubble core-dns with DNS proxy was enabled.
Some FQDN such as `registry.yarnpkg.com` does not work.

Destination protocol seems `ICMPv4`.
Normally, other FQDNs was allowed.

**Comment by gandro** [comment]
Thank you all for the detailed information. @bimmlerd and I started looking into this and here is our current finding:

The issue seems to occur when the DNS response is truncated - this causes most clients to re-try using TCP. DNS via TCP seems to be required to trigger this. What seems to be necessary to trigger the error is that Cilium has to deal with two TCP connections *using the same source port* in short sequence. This is also evident in the PCAPs where WireShark annotes the failing flow with `TCP Port numbers reused`.

What basically seems to happen is that Cilium's [shared DNS client](https://github.com/cilium/dns/blob/af788769786ae96ebbb5506625746bdf0a654cb6/shared_client.go) (a crucial piece in transparent DNS proxy mode)  implementation opens a socket for the given 5-tuple, performs a DNS request and response sequence, and then closes the socket again. If there is another DNS request re-using the same 5-tuple issues just afterwards, the `connect` call on the socket the second time fails presumably due to some socket state from the previous run still lingering. It's not clear yet if this lingering state is due to a race conditions in Ciliums code (we haven't found anything in that regard), Golang's asynchronous I/O runtime, or the kernel itself. What seems clear is that the window where "connect" fails after a previous "close" is very short, only a couple milliseconds. If the time between the two DNS requests reusing the same TCP ports is longer, then the issue is not reproducible.

We've also been able to reliably reproduce the issue on a modified Cilium version (that sleeps for a second before closing the socket) by invoking "dig -b "$local_addr#12345" +tcp one.one.one.one" from the client pod twice in a row. This strongly indicates that the problem really is related to quick succession TCP port-reuse - glibc and DNS truncation sufficient to achieve that, but not necessary.

We still have to find the root cause (Cilium, Go or kernel) in order to evaluate what the best fix would be.

**Comment by gandro** [comment]
We've done some further investigations, trying to figure out why the `connect()` call shortly after a prior `close` would fail with `EADDRNOTAVAIL` (aka `cannot assign requested address`). In particular, tried to create a minimal reproducer. We've been able to successfully reproduce the `EADDRNOTAVAIL` with a sequential C program that only uses `SO_REUSEADDR` - this seems to strongly suggest that it is the Linux kernel itself that prevents us from re-using the source port.

The exact reason why the kernel throws a `EADDRNOTAVAIL` even though we are using `SO_REUSEADDR` (which should allow source address reuse) remains unclear. What we have been able to establish is that the socket becomes available again either after a few hundred milliseconds or one minute - depending on the state of the TCP connection after `close`: If the socket enters `FIN_WAIT_1, it seems to reject new connect calls with `EADDRNOTAVAIL` for a few milliseconds, and then moves to `TIME_WAIT`, after which the `connect` call on the new socket succeeds. If the socket however enters `FIN_WAIT_2`, it also enters `TIME_WAIT` after a few hundred milliseconds, but the socket remains unavailable for the `connect` call even while in `TIME_WAIT`.

We are still investigating why the behavior differs for the two cases. The reason why one case is hit over the other other seems mainly based on the underlying network. For example, the QEMU network stack with SLiRP seems to reply with an `ACK` to the client's `FIN,ACK`, thus causing the client to enter `FIN_WAIT_2`. Where as bare metal machines in our testing seem to receive a `FIN,ACK` after sending their own `FIN,ACK`, causing to move from `FIN_WAIT_1` to `TIME_WAIT` directly instead.

A potential solution we've discussed is to prevent the shared client from closing the socket immediately, and instead keeping it around in case there are more incoming requests for the same source port. This has the result that Cilium would reuse the TCP connection for late requests. However, we are still investigating why the kernel reaches a state where the socket is unavailable for one minute, as the source port is basically unavailable in those cases once the TCP connection is closed.

**Comment by gandro** [comment]
> However, we are still investigating why the kernel reaches a state where the socket is unavailable for one minute, as the source port is basically unavailable in those cases once the TCP connection is closed.

We've established that the kernel fails [this condition here](https://github.com/torvalds/linux/blob/e9d22f7a6655941fc8b2b942ed354ec780936b3e/net/ipv4/tcp_ipv4.c#L155-L156) because `tw_ts_recent_stamp` is zero. Why it is zero, we haven't figured out yet. One way to avoid hitting the faulty `FIN_WAIT_2` state is to close the connection to upstream with a TCP RST (by using a `SO_LINGER` timeout of 0 seconds), but that means that Cilium DNS Proxy behaves very differently from the client:

We've also realized is that the TCP connection from the client (e.g. with glibc) to the proxy is persistent over all DNS requests for that source port - where as Cilium transparent DNS proxy wrongly terminates and reopens the TCP conncetion. This not only causes the `EADDRNOTAVAIL` some times (for the reasons mentioned above), but also seems to generally be slower than having the proxy persist the connection. So the best solution seems to be that we ought to teach Cilium DNS proxy to keep the connection to the upstream server open for as long as the connection between client and proxy is open too. Unfortunately, the DNS library we are using does not make this straight-forward to implement. We're investigating what the easiest solution might be.

**Comment by bimmlerd** [comment]
This is fixed in `main` and I've nominated it for backport to 1.16 - reopening until the backport has happened.

**Comment by bimmlerd** [comment]
Fixed in v1.16 in https://github.com/cilium/cilium/pull/34201
